interp/cubic.go

```diff
@@ -297,3 +296,122 @@ func fritschButlandEdgeDerivative(xs, ys, slopes []float64, leftEdge bool) float
 	}
 	return g
 }
+
+// fitWithSecondDerivatives fits a piecewise cubic predictor to (X, Y, d^2Y/dX^2) value
+// triples provided as three slices.
+// It panics if any of these is true:
+// - len(xs) < 2,
+// - elements of xs are not strictly increasing,
+// - len(xs) != len(ys),
+// - len(xs) != len(d2ydx2s).
+// Note that this method does not guarantee on its own the continuity of first derivatives.
+func (pc *PiecewiseCubic) fitWithSecondDerivatives(xs, ys, d2ydx2s []float64) {
+	n := len(xs)
+	if len(ys) != n {
+		panic(differentLengths)
+	}
+	if len(d2ydx2s) != n {
+		panic(differentLengths)
+	}
+	if n < 2 {
+		panic(tooFewPoints)
+	}
+	m := n - 1
+	pc.coeffs.Reset()
+	pc.coeffs.ReuseAs(m, 4)
+	for i := 0; i < m; i++ {
+		dx := xs[i+1] - xs[i]
+		if dx <= 0 {
+			panic(xsNotStrictlyIncreasing)
+		}
+		dy := ys[i+1] - ys[i]
+		dm := d2ydx2s[i+1] - d2ydx2s[i]
+		pc.coeffs.Set(i, 0, ys[i])                             // a_0
+		pc.coeffs.Set(i, 1, (dy-(d2ydx2s[i]+dm/3)*dx*dx/2)/dx) // a_1
+		pc.coeffs.Set(i, 2, d2ydx2s[i]/2)                      // a_2
+		pc.coeffs.Set(i, 3, dm/6/dx)                           // a_3
+	}
+	pc.xs = append(pc.xs[:0], xs...)
+	pc.lastY = ys[m]
+	lastDx := xs[m] - xs[m-1]
+	pc.lastDyDx = pc.coeffs.At(m-1, 1) + 2*pc.coeffs.At(m-1, 2)*lastDx + 3*pc.coeffs.At(m-1, 3)*lastDx*lastDx
+}
+
+// makeCubicSplineSecondDerivativeEquations generates the basic system of linear equations
+// which have to be satisfied by the second derivatives to make the first derivatives of a
+// cubic spline continuous. It panics if elements of xs are not strictly increasing, or
+// len(xs) != len(ys).
+// makeCubicSplineSecondDerivativeEquations returns a tri-diagonal matrix A and a vector b
+// defining a system of linear equations A*m = b for second derivatives vector m.
+func makeCubicSplineSecondDerivativeEquations(xs, ys []float64) (*mat.Tridiag, mat.MutableVector) {
+	n := len(xs)
+	if len(ys) != n {
+		panic(differentLengths)
+	}
+	b := make([]float64, n)
+	m := n - 1
+	// Diagonal of A:
+	d := make([]float64, n)
+	// Sub-diagonal of A:
+	dl := make([]float64, m)
+	// Super-diagonal of A:
+	du := make([]float64, m)
+	if n > 2 {
+		for i := 0; i < m; i++ {
+			dx := xs[i+1] - xs[i]
+			if dx <= 0 {
+				panic(xsNotStrictlyIncreasing)
+			}
+			slope := (ys[i+1] - ys[i]) / dx
+			if i > 0 {
+				b[i] += slope
+				d[i] += dx / 3
+				du[i] = dx / 6
+			}
+			if i < m-1 {
+				b[i+1] -= slope
+				d[i+1] += dx / 3
+				dl[i] = dx / 6
+			}
+		}
+	}
+	return mat.NewTridiag(n, dl, d, du), mat.NewVecDense(n, b)
```

Comment:
There should be a ticket to use `SymTridiag` when it exists and has `SolveVecTo`. You could already use `SymBandDense` and `BandCholesky`.

Reply:
I think the matrix isn't symmetric in this case.

Reply:
Ok, it's not symmetric because of the boundary conditions. Currently there is only `y''(left) = y''(right) = 0` and in this case the matrix can be made symmetric by eliminating the first and last unknown. I don't know what other BCs you plan and whether it's worth the effort.

Reply:
I plan to add those listed in https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.CubicSpline.html and custom ones (`a*y'(left) + b*y''(left) == c && d*y'(right) + e*y''(right) == f`). Started with natural just to get the overall design right without having to redo a lot of work.

Reply:
Ok.